docker/plugin/digitile-monstache-plugin.go

```diff
@@ -0,0 +1,459 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/rwynn/monstache/monstachemap"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"log"
+	"os"
+	"strings"
+	"sync"
+)
+
+var warnLog = log.New(os.Stdout, "WARN ", log.Flags())
+
+var assetFields = map[string]bool{
+	"_id":             true,
+	"org":             true,
+	"filename":        true,
+	"filetype":        true,
+	"content":         true,
+	"tags":            true,
+	"thumbnail_url":   true,
+	"remote":          true,
+	"users":           true,
+	"created":         true,
+	"remote_modified": true,
+	"owner":           true,
+	"editors":         true,
+	"counters":        true,
+}
+
+var searchUpdateFields = map[string]bool{
+	"filename":        true,
+	"content":         true,
+	"tags":            true,
+	"thumbnail_url":   true,
+	"users":           true,
+	"remote_modified": true,
+	"editors":         true,
+	"counters":        true,
+	"forceSync":       true,
+}
+
+const DELIMITER = "|"
+
+func Map(input *monstachemap.MapperPluginInput) (output *monstachemap.MapperPluginOutput, err error) {
+
+	if input.Operation == "i" {
+
+		output, err = ProcessDocument(input)
+
+	} else {
+
+		if SearchFieldsUpdated(input.UpdateDescription) {
+
+			output, err = ProcessDocument(input)
+
+		} else {
+			output = &monstachemap.MapperPluginOutput{
+				Skip: true,
+			}
+		}
+	}
+
+	return output, err
+}
+
+func ProcessDocument(input *monstachemap.MapperPluginInput) (output *monstachemap.MapperPluginOutput, err error) {
+
+	//start := time.Now()
+
+	document := input.Document
+	assetId := document["_id"].(primitive.ObjectID).Hex()
+
+	defer func() {
+		if recoveredError := recover(); recoveredError != nil {
+			err = fmt.Errorf("Syncing asset with ID %s failed with an unknow error: %s",
+				assetId, recoveredError)
+		}
+	}()
+
+	output = &monstachemap.MapperPluginOutput{
+		Routing:  GetRouting(document),
+		Document: document,
+	}
+
+	PruneDocument(document)
+	TransformDocument(document)
+
+	var wg sync.WaitGroup
+	lookupErrors := make(chan string, 3)
+	documentFragments := make(chan map[string]interface{}, 3)
+
+	wg.Add(1)
+	go AddFileTypeFields(input, document, &wg, documentFragments, lookupErrors)
+	wg.Add(1)
+	go AddTagNames(input, document, &wg, documentFragments, lookupErrors)
+	wg.Add(1)
+	go AddUserDetails(input, document, &wg, documentFragments, lookupErrors)
+
+	wg.Wait()
+	close(lookupErrors)
+	close(documentFragments)
```

Comment:
Synchronizes all Mongo lookups using a [WaitGroup](https://golang.org/pkg/sync/#WaitGroup) where all lookups are executed in parallel and the code waits for all 3 [Goroutines](https://tour.golang.org/concurrency/1) to complete before handling errors and results. The results are communicated back through Golang [channels](https://tour.golang.org/concurrency/2). The original document is not modified since concurrent modifications are not allowed in Golang maps. Instead the individual document fragments are sent back and added to the document after all Goroutines are completed.

Reply:
Why are using 3 as a concurrency? Could we make this section abstract so it can be configured from the config file/environment?

Reply:
This is the plugin code not Monstache. The plugin only has 3 lookups that need to be done in parallel: tags, filetype and users